google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java

```diff
@@ -33,6 +103,348 @@
    */
   @Nullable
   CloudProject showDialog(@Nullable CloudProject cloudProject) {
-    return cloudProject;
+    // create dialog wrapper for the form once, create services and subscribe to login changes once.
+    if (dialogWrapper == null) {
+      dialogWrapper = new ProjectSelectionDialogWrapper();
+
+      dialogWrapper.setTitle(GctBundle.getString("cloud.project.selector.dialog.title"));
+      // disabled unless project is selected in the list.
+      dialogWrapper.setOKActionEnabled(false);
+      // install additional IDEA UI.
+      installTableSpeedSearch(projectListTable);
+
+      Stream.of(ProjectManager.getInstance().getOpenProjects())
+          .forEach(
+              project ->
+                  project
+                      .getMessageBus()
+                      .connect(dialogWrapper.getDisposable() /* disconnect once dialog is gone. */)
+                      .subscribe(
+                          GoogleLoginListener.GOOGLE_LOGIN_LISTENER_TOPIC,
+                          this::loadUsersAndProjects));
+    }
+
+    loadUsersAndProjects();
+    setSelectedProject(cloudProject);
+
+    DialogManager.show(dialogWrapper);
+
+    int result = dialogWrapper.getExitCode();
+    return result == DialogWrapper.OK_EXIT_CODE ? getSelectedProject() : null;
+  }
+
+  @VisibleForTesting
+  void setSelectedProject(CloudProject cloudProject) {
+    if (cloudProject != null && !Strings.isNullOrEmpty(cloudProject.googleUsername())) {
+      Optional<CredentialedUser> loggedInUser =
+          Services.getLoginService().getLoggedInUser(cloudProject.googleUsername());
+      if (loggedInUser.isPresent()) {
+        selectedProjectsByAccount.put(loggedInUser.get(), cloudProject.projectName());
+        accountComboBox.setSelectedItem(loggedInUser.get());
+      } else {
+        // specified user is not in logged in user list, clear the account selection.
+        accountComboBox.setSelectedItem(null);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  CloudProject getSelectedProject() {
+    CredentialedUser user = accountComboBox.getItemAt(accountComboBox.getSelectedIndex());
+    return CloudProject.create(getSelectedProjectName(), user.getEmail());
+  }
+
+  @VisibleForTesting
+  void loadUsersAndProjects() {
+    Collection<CredentialedUser> credentialedUsers =
+        Services.getLoginService().getAllUsers().values();
+    if (credentialedUsers.isEmpty()) {
+      showSignInRequest();
+    } else {
+      hideSignInRequest();
+      cachedProjectList.clear();
+      accountComboBox.removeAllItems();
+      for (CredentialedUser user : credentialedUsers) {
+        accountComboBox.addItem(user);
+        updateProjectList(user);
```

Comment:
This is what I was thinking, lmkwyt

- `loadAllProjects` is responsible for loading and caching the projects for all users. Thats is; it doesnt update the UI.
- `setSelectedProject` then updates the dropdown with the current CloudProject, and in turn, causes the table model to update (pulling the list of projects from the cache).
- the listener on the account combobox updates the table model like in the above bullet

Its similar, but I think this will separate concerns a bit more cleanly. Sorry if I am contradicting previous things I've said with these comments. The motivation is I think that updateProjectList is a bit too complicated and doing too much (loading projects, updating the UI..).

Reply:
sounds reasonable, I'll do a simple refactor of this idea, trying to clean up `updateProjectList' a bit more